### PR TITLE
Add ARM builds

### PR DESCRIPTION
**Pull request by pandadefi**
Building on the already started changes from https://github.com/hexpm/bob/pull/152 I have taken into consideration the comments you have made @ericmj please have a look and tell me how to further continue.

### CODE REVIEW

File: lib/bob/job/build_otp.ex [hunk 1]
```diff
@@ -1,10 +1,10 @@
 defmodule Bob.Job.BuildOTP do
   require Logger
 
-  def run(ref_name, ref, linux) do
+  def run(ref_name, ref, linux, arch) do
```
Comment (ericmj) [inline lib/bob/job/build_otp.ex:4]: `Bob.Queue.add({Bob.Job.BuildOTP, arch}, [ref_name, ref, linux])` will add `arch` as the first argument to the job.

```suggestion
  def run(arch, ref_name, ref, linux) do
```

File: lib/bob/job/otp_checker.ex [hunk 2]
```diff
@@ -1,28 +1,31 @@
...
   @linuxes ["ubuntu-18.04", "ubuntu-20.04", "ubuntu-22.04"]
+  @arches ["amd64", "arm64"]
 
   def run(_type) do
     for linux <- @linuxes,
-        {ref_name, ref} <- Bob.GitHub.diff(@repo, "builds/otp/#{linux}"),
-        build_ref?(linux, ref_name),
-        do: Bob.Queue.add(Bob.Job.BuildOTP, [ref_name, ref, linux])
+        arch <- @arches,
+        {ref_name, ref} <- Bob.GitHub.diff(@repo, "builds/otp/#{linux}-#{arch}"),
```
Comment (ericmj) [inline lib/bob/job/otp_checker.ex:9]: Let's use this paths to make it easier to do path rewrites on the CDN.

```suggestion
        {ref_name, ref} <- Bob.GitHub.diff(@repo, "builds/otp/#{arch}/#{linux}"),
```

File: priv/scripts/otp/otp.sh [hunk 3]
```diff
@@ -5,12 +5,21 @@ set -euox pipefail
...
+echo "Building $1 $2 $3 $4"
+
+if [[ "$arch" == "arm64" ]]; then
+  container="otp-build-${linux}-${ref_name}-arm64"
+elif [[ "$arch" == "amd64" ]]; then
+  container="otp-build-${linux}-${ref_name}-amd64"
+else
+  echo "Invalid architecture. Please specify either 'arm64' or 'amd64'."
+  exit 1
+fi
```
Comment (ericmj) [inline priv/scripts/otp/otp.sh:21]: Different container names are not needed as the builds will run on different machines.

```suggestion
```

File: priv/scripts/otp/otp.sh [hunk 4]
```diff
@@ -30,21 +39,22 @@ docker build \
...
+
 docker rm -f ${container} || true
 docker run -t -e OTP_REF=${ref_name} --name=${container} ${image}:${tag}
 
 docker cp ${container}:/home/build/out/${ref_name}.tar.gz ${ref_name}.tar.gz
 
 docker rm -f ${container}
 
-aws s3 cp ${ref_name}.tar.gz s3://s3.hex.pm/builds/otp/${linux}/${ref_name}.tar.gz --cache-control "public,max-age=3600" --metadata "{\"surrogate-key\":\"builds builds/otp builds/otp/${linux} builds/otp/${linux}/${ref_name}\",\"surrogate-control\":\"public,max-age=604800\"}"
+aws s3 cp ${ref_name}.tar.gz s3://s3.hex.pm/builds/otp/${linux}-${arch}/${ref_name}.tar.gz --cache-control "public,max-age=3600" --metadata "{\"surrogate-key\":\"builds builds/otp builds/otp/${linux} builds/otp/${linux}/${ref_name}\",\"surrogate-control\":\"public,max-age=604800\"}"
```
Comment (ericmj) [inline priv/scripts/otp/otp.sh:50]:
```suggestion
aws s3 cp ${ref_name}.tar.gz s3://s3.hex.pm/builds/otp/${arch}/${linux}/${ref_name}.tar.gz --cache-control "public,max-age=3600" --metadata "{\"surrogate-key\":\"builds builds/otp builds/otp/${arch} builds/otp/${arch}/${linux} builds/otp/${arch}/${linux}/${ref_name}\",\"surrogate-control\":\"public,max-age=604800\"}"
```

File: priv/scripts/otp/otp.sh [hunk 5]
```diff
@@ -30,21 +39,22 @@ docker build \
...
 
-aws s3 cp s3://s3.hex.pm/builds/otp/${linux}/builds.txt builds.txt || true
+aws s3 cp s3://s3.hex.pm/builds/otp/${linux}-${arch}/builds.txt builds.txt || true
```
Comment (ericmj) [inline priv/scripts/otp/otp.sh:52]:
```suggestion
aws s3 cp s3://s3.hex.pm/builds/otp/${arch}/${linux}/builds.txt builds.txt || true
```

File: priv/scripts/otp/otp.sh [hunk 6]
```diff
@@ -30,21 +39,22 @@ docker build \
...
 touch builds.txt
 sed -i "/^${ref_name} /d" builds.txt
 echo -e "${ref_name} ${ref} $(date -u '+%Y-%m-%dT%H:%M:%SZ')\n$(cat builds.txt)" > builds.txt
 sort -u -k1,1 -o builds.txt builds.txt
-aws s3 cp builds.txt s3://s3.hex.pm/builds/otp/${linux}/builds.txt --cache-control "public,max-age=3600" --metadata "{\"surrogate-key\":\"builds builds/otp builds/otp/${linux} builds/otp/${linux}/txt\",\"surrogate-control\":\"public,max-age=604800\"}"
+aws s3 cp builds.txt s3://s3.hex.pm/builds/otp/${linux}-${arch}/builds.txt --cache-control "public,max-age=3600" --metadata "{\"surrogate-key\":\"builds builds/otp builds/otp/${linux} builds/otp/${linux}/txt\",\"surrogate-control\":\"public,max-age=604800\"}"
```
Comment (ericmj) [inline priv/scripts/otp/otp.sh:57]:
```suggestion
aws s3 cp builds.txt s3://s3.hex.pm/builds/otp/${arch}/${linux}/builds.txt --cache-control "public,max-age=3600" --metadata "{\"surrogate-key\":\"builds builds/otp builds/otp/${arch} builds/otp/${arch}/${linux} builds/otp/${arch}/${linux}/txt\",\"surrogate-control\":\"public,max-age=604800\"}"
```

File: priv/scripts/otp/otp.sh [hunk 7]
```diff
@@ -30,21 +39,22 @@ docker build \
...
 
-fastly_purge $BOB_FASTLY_SERVICE_HEXPM "builds/otp/${linux}/txt builds/otp/${linux}/${ref_name}"
-fastly_purge $BOB_FASTLY_SERVICE_BUILDS "builds/otp/${linux}/txt builds/otp/${linux}/${ref_name}"
+fastly_purge $BOB_FASTLY_SERVICE_HEXPM "builds/otp/${linux}-${arch}/txt builds/otp/${linux}-${arch}/${ref_name}"
+fastly_purge $BOB_FASTLY_SERVICE_BUILDS "builds/otp/${linux}-${arch}/txt builds/otp/${linux}-${arch}/${ref_name}"
```
Comment (ericmj) [inline priv/scripts/otp/otp.sh:60]:
```suggestion
fastly_purge $BOB_FASTLY_SERVICE_HEXPM "builds/otp/${arch}/${linux}/txt builds/otp/${arch}/${linux}/${ref_name}"
fastly_purge $BOB_FASTLY_SERVICE_BUILDS "builds/otp/${arch}/${linux}/txt builds/otp/${arch}/${linux}/${ref_name}"
```

**Review by ericmj — COMMENTED** [review]
Sorry for missing this PR. It looks good other than some minor changes.

I will take this and do some testing. If any major changes are needed I will let you know.

**Comment by pandadefi** [comment]
Great

**Comment by ericmj** [comment]
Huge thanks @acco and @pandadefi!

**Comment by ericmj** [comment]
The builds have started to run, will probably take about a day to run through all the builds for ubuntu-20.04, ubuntu-22.04, and ubuntu-24.04.

**Pull request merged**